### PR TITLE
[SYCLomatic] Enable data sampling in CodePin

### PR DESCRIPTION
**Pull request by AndyCHHuang**

### CODE REVIEW

**Review by ziranzha — APPROVED** [review]

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 1]
```diff
@@ -23,14 +23,27 @@
...
 #include <utility>
 #include <vector>
+#include <stdlib.h>
+
+#ifndef CODEPIN_RAND_SEED
+#define CODEPIN_RAND_SEED 0
+#endif
+
+#ifndef CODEPIN_SAMPLING_THRESHHOLD
+#define CODEPIN_SAMPLING_THRESHHOLD 20
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]: Please add comments on the 3 macros on:  What it is, When to use, how to use;

Comment (zhimingwang36) [reply clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]: 2. how about:  emit build time warnning message to notify the default samping is enabled, and how to change the default sampling?

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 2]
```diff
@@ -23,14 +23,27 @@
...
+#endif
+
+#ifndef CODEPIN_SAMPLING_RATE
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]: codepin-report.py should also contain the sampling infomation, and tells user how to change the sampling info.

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 3]
```diff
@@ -212,6 +235,10 @@ template <class... Args>
 void gen_log_API_CP(const std::string &cp_id, std::string device_name,
                     size_t free_byte, size_t total_byte, float elapse_time,
                     queue_t queue, Args... args) {
+  if (!rand_seed_setup) {
+    srand(CODEPIN_RAND_SEED);
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp:265]: how about:  +  print runtime info on CodePin info (like sampling info).

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 4]
```diff
@@ -23,22 +23,58 @@
...
+#define CODEPIN_RAND_SEED 0
+#endif
+
+// Data size thresh hold for data sampling.
+// array/pointer size larger than the threshhold will be sampled.
+#ifndef CODEPIN_SAMPLING_THRESHHOLD
+#define CODEPIN_SAMPLING_THRESHHOLD 20
+#endif
+
+// Sampling rate, interval: (0, 1]
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]:
```suggestion
// Sampling percent, interval: [0, 100]
// 0: No data will be logged.  
// 100: all data will be logged.
```

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 5]
```diff
@@ -23,22 +23,58 @@
...
 #include <vector>
+#include <iostream>
+#include <stdlib.h>
+
+// Random seed for data sampling.
+#ifndef CODEPIN_RAND_SEED
+#define CODEPIN_RAND_SEED 0
+#endif
+
+// Data size thresh hold for data sampling.
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]:
```suggestion
// Data size threshold to trigger data sampling.
```

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 6]
```diff
@@ -23,22 +23,58 @@
...
+// array/pointer size larger than the threshhold will be sampled.
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]: ### 
```suggestion
// array/pointer size larger than the threshold will be sampled.
```

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 7]
```diff
@@ -23,22 +23,58 @@
...
+#endif
+
+// Sampling rate, interval: (0, 1]
+#ifndef CODEPIN_SAMPLING_RATE
+#define CODEPIN_SAMPLING_RATE 0.01
+#endif
+
+#define CODEPIN_TO_STR(x) CODEPIN_STR(x)
+#define CODEPIN_STR(x) #x
+#pragma message("CodePin sampling feature is enabled.")
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]:
```suggestion
#pragma message("CodePin sampling feature is enabled for data dump. As follow list 3 configs for data sampling:")
```

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 8]
```diff
@@ -23,22 +23,58 @@
...
+#pragma message("CODEPIN_RAND_SEED: " CODEPIN_TO_STR(CODEPIN_RAND_SEED))
+#pragma message("CODEPIN_SAMPLING_THRESHHOLD: " CODEPIN_TO_STR(                \
+    CODEPIN_SAMPLING_THRESHHOLD))
+#pragma message("CODEPIN_SAMPLING_RATE: " CODEPIN_TO_STR(CODEPIN_SAMPLING_RATE))
+#pragma message(                                                               \
+    "Define the macros in the build command to change sampling config.")
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]:
```suggestion
    "Define the macros in the build command to change sampling configs. Also refer to codepin.hpp for definitions and default value of the macros. ")
```

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 9]
```diff
@@ -212,6 +258,15 @@ template <class... Args>
 void gen_log_API_CP(const std::string &cp_id, std::string device_name,
                     size_t free_byte, size_t total_byte, float elapse_time,
                     queue_t queue, Args... args) {
+  if (!rand_seed_setup) {
+    srand(CODEPIN_RAND_SEED);
+    rand_seed_setup = true;
+    std::cout << "CodePin Sampling is enabled:" << std::endl;
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]:
```suggestion
    std::cout << "CodePin Sampling is enabled for data dump As follow list 3 configs for data sampling:" << std::endl;
```

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**